core/src/main/java/com/graphhopper/routing/profiles/EncodedValueFactory.java

```diff
@@ -23,4 +23,6 @@
      * of it.
      */
     EncodedValue create(String encodedValueString);
+
+    Class<? extends Enum> findValues(String name);
```

Comment:
Intuitively I would take the encoding manager (or encoded value lookup), ask for the encoded value with a certain name and then ask that (enum) encoded value for its possible values. This seems more obvious to me than going back to the factory (where the encoded values were produced) and ask which possible values were baked into the encoded value when it was created. If I was to implement my own `EncodedValueFactory` I rather only deal with the creation of the encoded values than also implementing this `findValues` method (it has to match the implementation of `create` which seems a bit ugly). In any case we should document (and maybe rename) `findValues`. I particularly find the 'find' in `findValues` misleading, because this sounds like we are trying to 'find' something. But this is exactly not the case, because  (compared to `EncodingManager`) the `EncodedValueFactory` is not the place *holding* the encoded values. Maybe something like:

```java
/**
 * Returns the set of possible values of the EnumEncodedValue with the given name.
 * When implementing this method make sure to accept any name for which `create` creates an `EnumEncodedValue`.
**/
Class<? extends Enum> getEnumValues(String encodedValueName)
```

But like I said I would prefer moving this functionality into `EncodingManager`/`EnumEncodedValue` somehow.

Reply:
This is not that much about values of an encoded value (like motorway or primary -> this is done via RoadAccess.values()) it is more a reflection replacement: we have a name in the script like `road_class` and want the associated `RoadClass` enum for it. I could have done this via Java reflections but I do not like it much (and is hard to port to Android or iOS).

> (it has to match the implementation of create which seems a bit ugly).

Yeah, but I did not find something better yet

Reply:
> This is not that much about values of an encoded value 

Hm is it not? For example in InfoResource we do:

```java
Class<? extends Enum> enumClass = evFactory.findValues(encodedValue);
for (Object o : enumClass.getEnumConstants()) {
       possibleValueList.add(o.toString());
}
```

We seem to be interested in getting the possible values (ok the Enum class, but that is almost the same) of an encoded (enum) value. And we are using the ev*factory* for this. But to me it would seem much more obvious to use the encoded value (and the thing that stores it (=encoding manager)) for this. For example its also strange to me that we need to pass the encoded value factory to `CustomWeighting` (for the same reason).

Also think about this: If we run the import for a graph and store everything in a 'graph-database' like we envisioned earlier we want to be able to get all the information from this after loading from disk. We do not want to be forced to create the same EVFactory just so it tells us what kind of enum type our encoded values were created with. This information should be part of the encoded values IMO.

> Yeah, but I did not find something better yet

Ok I have not looked into this and I am not saying its easy and also I rather avoid using reflection, but conceptionally I am saying that it should be possible to get the values of an enum encoded value by just looking at it without using the factory that *created* the encoded value.

Reply:
The problem is that the `EnumEncodedValue` does not know its values, only the enum itself does and we need a method that gives us this from the string (i.e. reflection or this). Moving the method into EncodingManager could make sense too, sure.

Reply:
> Moving the method into EncodingManager could make sense too, sure.

Shall I try this? That would make for a better API already IMO. And maybe later we can still make the `EnumEncodedValue` 'know' its values or does that not make sense?

Reply:
> Shall I try this? 

Yes, would be nice!

> And maybe later we can still make the EnumEncodedValue 'know' its values or does that not make sense?

It would be kind of duplicated information as this is already in the specific Enum. But in general it could make sense to move away from enums and towards strings (has also several disadvantages) but then we could make it working that in the future we can read graphs from older graphhopper versions.

Reply:
> private final Class<E> enumType;  // added

Indeed, sounds good!

Reply:
Oh, you mean we could we just return the array `arr` instead of the enum lookup.

Reply:
Yeah exactly, arr is basically all we need and we have it in `EnumEncodedValue` already...